File: src/main/java/com/binhjcao/graveler/action/Fall.java
package com.binhjcao.graveler.action;

import com.binhjcao.graveler.util.PointedWorld;
import com.binhjcao.graveler.util.WorldExtensions;
import net.minecraft.core.BlockPos;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.world.entity.item.FallingBlockEntity;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.Blocks;
import net.minecraft.core.Direction;

public record Fall(BlockPos pos) implements Action {
  @Override
  public double weight() {
    return 1 / 128.0;
  }

  @Override
  public void apply(ActionContext context) {
    fallAt(context.world(), pos);
  }

  private static void fallAt(ServerLevel level, BlockPos pos) {
    if (!level.hasChunksAt(
        pos.getX() - 32,
        pos.getY() - 32,
        pos.getZ() - 32,
        pos.getX() + 32,
        pos.getY() + 32,
        pos.getZ() + 32)) {
      forceInstantFallAt(level, pos);
      return;
    }

    if (level.isClientSide()) {
      return;
    }

    forceFallAt(level, pos);
  }

  private static void forceFallAt(Level level, BlockPos pos) {
    FallingBlockEntity.fall(level, pos, level.getBlockState(pos));
  }

  private static void forceInstantFallAt(Level level, BlockPos pos) {
    var state = level.getBlockState(pos);
    level.setBlockAndUpdate(pos, Blocks.AIR.defaultBlockState());

    PointedWorld below = WorldExtensions.pointedAt(level, pos).move(Direction.DOWN);
    while (below.allowsFallThrough() && below.pos().getY() > 0) {
      below = below.move(Direction.DOWN);
    }

    if (below.pos().getY() > 0) {
      level.setBlockAndUpdate(below.pos().above(), state);
    }
  }
}

File: src/main/kotlin/graveler/action/Fall.kt
package graveler.action

import graveler.util.PointedWorld
import graveler.util.pointedAt
import net.minecraft.block.Blocks
import net.minecraft.block.CauldronBlock
import net.minecraft.entity.item.FallingBlockEntity
import net.minecraft.util.Direction
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World

data class Fall(val pos: BlockPos) : Action {
  override val weight: Double = 1 / 128.0

  override fun apply(context: ActionContext) {
    context.world.fallAt(pos)
  }

  private fun World.fallAt(pos: BlockPos) {
    if (!isAreaLoaded(pos, 32)) {
      forceInstantFallAt(pos)
      return
    }

    if (isRemote) {
      return
    }

    forceFallAt(pos)
  }

  private fun World.forceFallAt(pos: BlockPos) {
    addEntity(
      FallingBlockEntity(
        this,
        pos.x + 0.5,
        pos.y.toDouble(),
        pos.z + 0.5,
        getBlockState(pos)))
  }

  private fun World.forceInstantFallAt(pos: BlockPos) {
    val state = getBlockState(pos)

    setBlockState(pos, Blocks.AIR.defaultState)

    var below = this.pointedAt(pos).move(Direction.DOWN)
    while (below.allowsFallThrough && below.pos.y > 0) {
      below = below.move(Direction.DOWN)
    }

    if (below.pos.y > 0) {
      // Forge: Fix loss of state information during world gen.
      setBlockState(below.pos.up(), state)
    }
  }

  private val PointedWorld.allowsFallThrough: Boolean
    get() = (!material.blocksMovement() || material.isLiquid) &&
      block !is CauldronBlock
}
